src/graphic/animation/spritesheet_animation.cc

```diff
@@ -0,0 +1,222 @@
+		}
+	}
+
+	// Frame width and height
+	w = sheet->width() / columns;
+	h = sheet->height() / rows;
+}
+
+void SpriteSheetAnimation::SpriteSheetMipMapEntry::blit(uint32_t idx,
+                                           const Rectf& source_rect,
+                                           const Rectf& destination_rect,
+                                           const RGBColor* clr,
+                                           Surface* target) const {
+	assert(sheet != nullptr);
+	assert(target);
+	assert(static_cast<int>(idx) <= columns * rows);
+
+	const int column = idx % columns;
+	const int row = idx / columns;
+
+	Rectf frame_rect(source_rect.x + column * width(), source_rect.y + row * height(),
+					 source_rect.w, source_rect.h);
+
+	if (!has_playercolor_masks || clr == nullptr) {
+		target->blit(destination_rect, *sheet, frame_rect, 1., BlendMode::UseAlpha);
+	} else {
+		assert(playercolor_mask_sheet != nullptr);
+		target->blit_blended(
+		   destination_rect, *sheet, *playercolor_mask_sheet, frame_rect, *clr);
+	}
+}
+
+int SpriteSheetAnimation::SpriteSheetMipMapEntry::width() const {
+	return w;
+}
+int SpriteSheetAnimation::SpriteSheetMipMapEntry::height() const {
+	return h;
+}
+
+/*
+==============================================================================
+
+SpriteSheetAnimation IMPLEMENTATION
+
+==============================================================================
+*/
+
+SpriteSheetAnimation::SpriteSheetAnimation(const LuaTable& table, const std::string& basename)
+   : Animation(table) {
+	try {
+		// Get image files
+		if (basename.empty() || !table.has_key("directory")) {
+			throw Widelands::GameDataError(
+			   "Animation did not define both a basename and a directory for its sprite sheet file");
+		}
+		const std::string directory = table.get_string("directory");
+
+		// Frames, rows and columns
+		nr_frames_ = table.get_int("frames");
+		rows_ = table.get_int("rows");
+		columns_ = table.get_int("columns");
+
+		add_available_scales(basename, directory);
+
+		// Perform some checks to make sure that the data is complete and consistent
+		const SpriteSheetMipMapEntry& first = dynamic_cast<const SpriteSheetMipMapEntry&>(*mipmaps_.begin()->second.get());
+		if (table.has_key("fps") && nr_frames_ == 1) {
+				throw Widelands::GameDataError("'%s' sprite sheet animation with one frame must not have 'fps'",
+				                               basename.c_str());
+		}
+
+		if (representative_frame() < 0 || representative_frame() > nr_frames_ - 1) {
+			throw Widelands::GameDataError("Animation has %d as its representative frame, but the frame indices "
+			                 "available are 0 - %d",
+			                 representative_frame(), nr_frames_ - 1);
+		}
+
+		if (rows_ * columns_ < nr_frames_) {
+			throw Widelands::GameDataError("Animation has %d frames, which does not fit into %d rows x %d columns",
+										   nr_frames_, rows_, columns_);
+		}
```

Comment:
Also throw an error if we have blank rows?

Reply:
Done.